incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -31,44 +39,176 @@ type HierarchicalNamespaceReconciler struct {
-	// instances on apiserver accordingly.
-	reason := "new/updated hierarchical namespace"
-	r.hcr.enqueueAffected(log, reason, nm)
+		// Set the "Owner" in the forest of the hierarchical namespace to the current namespace.
+		log.Info("Setting the subnamespace's owner in the forest", "owner", pnm, "namespace", nm)
+		// TODO rename RequiredChildOf to Owner in the forest. See issue:
+		//  https://github.com/kubernetes-sigs/multi-tenancy/issues/469
+		ns.RequiredChildOf = pnm
 
-	// TODO sync with forest to report conflicts in hns.Status.State. See issue:
-	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/487
+		// Enqueue the not-yet existent self-serve subnamespace. The HierarchyConfig reconciler will
+		// create the namespace and the HierarchyConfig instances on apiserver accordingly.
+		reason := "new/updated hierarchical namespace"
+		r.hcr.enqueueAffected(log, reason, nm)
 
-	return ctrl.Result{}, nil
+		inst.Status.State = api.Missing
+		return
+	}
+
+	log.Info("The self-serve subnamespace already exists", "namespace", nsInst.Name)
+	// If the namespace is not a self-serve namespace or it's a self-serve namespace of another
+	// namespace. Report the conflict.
+	if nsInst.Annotations[api.AnnotationOwner] != pnm {
+		log.Info("The namespace cannot be used as a self-serve subnamespace")
+		inst.Status.State = api.Conflict
+		return
+	}
+
+	log.Info("The namespace has the correct owner annotation", "annotation", nsInst.Annotations[api.AnnotationOwner])
+	switch hcInst.Spec.Parent {
+	case "":
+		log.Info("Parent is not set", "namespace", nm)
+		// This case is rare. It means the namespace is created with the right annotation but
+		// no HC. This could be a transient state before HC reconciler finishes creating the HC
+		// or a human manually created the namespace with the right annotation but no HC.
+		// Both cases meant to create this namespace as HNS.
+		log.Info("Setting the subnamespace's owner in the forest", "owner", pnm, "namespace", nm)
+		ns.RequiredChildOf = pnm
+
+		reason := "new/updated hierarchical namespace"
+		r.hcr.enqueueAffected(log, reason, nm)
+
+		// We will set it as "Conflict" though it's just a short transient state. Once the hc is
+		// reconciled, this HNS will be enqueued and then set the state to "Ok".
+		inst.Status.State = api.Conflict
+	case pnm:
+		log.Info("Setting the HierarchicalNamespace state to Ok")
+		inst.Status.State = api.Ok
+	default:
+		log.Info("Self-serve subnamespace is already owned by another parent", "child", nm, "otherParent", hcInst.Spec.Parent)
+		inst.Status.State = api.Conflict
+	}
+}
+
+// enqueueAffected enqueues all affected hierarchicalNamespace instances for later reconciliation. This occurs
+// in a goroutine so the caller doesn't block; since the reconciler is never garbage-collected, this is safe.
+func (r *HierarchicalNamespaceReconciler) enqueueAffected(log logr.Logger, nm, pnm, reason string) {
```

Comment:
Maybe instead of `enqueueAffected`, just say `enqueue`? This only enqueues a single HNS object, it doesn't go looking for other stuff (like the other `enqueueAffected` methods).